worker/run.py

```diff
@@ -49,56 +51,100 @@ def main():
         args.websocket_base_url = urllib.parse.urlunparse(sync_url)
 
     if args.reload:
+        logging.info("Reload enabled")
         path = Path(__file__).parent
 
-        p, event = run_sync_in_process(args)
-        for _ in watch(path):
+        stop_watching_event = asyncio.Event()
+        p = run_sync_in_process(stop_watching_event, args)
+
+        def watch_filter(change, path: str):
+            # do not reload when models are downloaded
+            return not path.startswith(settings.MODELS_DIR.absolute().as_posix())
+
+        for _ in watch(path, watch_filter=watch_filter, stop_event=stop_watching_event):
             logging.info("Source code change detected, reloading worker")
-            event.set()
+            p.terminate()
             p.join()
-            p, event = run_sync_in_process(args)
+            p = run_sync_in_process(stop_watching_event, args)
 
     else:
-        run_sync(args, Event())
+        run_sync(args)
 
 
-def run_sync_in_process(args):
-    event = Event()
-    p = Process(target=run_sync, args=(args, event))
+def run_sync_in_process(stop_watching_event: asyncio.Event, args):
+    p = Process(target=run_sync, args=(args,))
     p.start()
-    return p, event
+
+    def handle_signal(sig, *args):
+        if p.pid is None:
+            raise Exception("Process has no PID")
+
+        os.kill(p.pid, sig)
```

Comment:
what does this do?

Reply:
it passes the signal from the reload watcher process to the actual worker process

Reply:
can I have comments explaining this?

Reply:
done